proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,351 @@
+# Dark Mode Extension Icons
+
+**Summary**
+
+Feature to enable developers to enhance extension icon visibility in dark mode.
+
+**Document Metadata**
+
+**Author:** &lt;solomonkinard>
+
+**Sponsoring Browser:** Chromium.
+
+**Contributors:** &lt;oliverdunk>
+
+**Created:** 2024-04-05
+
+**Related Issues:**
+* https://crbug.com/893175
+* https://github.com/w3c/webextensions/issues/229
+
+## Motivation
+
+### Objective
+
+Extension developers will be able to supply and define a set of icons to be used
+in the event that user has expressed the preference for a page that has a dark
+theme. The set of dark mode icons can and will be made automatically visible
+once the system setting is in dark mode, thereby enhancing the overall browser
+experience.
+
+#### Use Cases
+
+1. Improved icon visibility in the extension toolbar.
+1. Improved icon visibility on management the management and shortcuts pages.
+1. Dark mode icon declarations made possible through the extension manifest.
+1. setIcon() will allow setting of dark and/or light mode specific icons.
+1. Improved icon visibility on context menu (Chrome and Safari reply on default
+icons, only Firefox can specify icons).
+1. Improved icon visibility on side panel (Chrome replies on default icons,
+Firefox can specify icons).
+
+### Known Consumers
+
+The Chromium bug has a significant amount of developer interest.
+
+## Specification
+
+### Schema
+
+// TODO: Obtain consensus on any of the following options before merging.
+
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+---
+
+#### Option (donut option)
+
+Don't implement setIcon() for dark mode, as it's used for dynamic icons anyway.
+
+* Pros: Easy to implement.
+* Cons: Developers would need to find their own ways of detecting dark mode.
```

Comment:
If this proposal doesn't support `action.setIcon()` for dark and light icons. Can browsers provide APIs for detecting the browser (not OS) dark/light theme and a onChangeListener? For example:

```
browser.runtime.getBrowserTheme() // return 'dark' or 'light'

browser.runtime.onBrowserThemeChanged.addListener((newTheme) => {
    // newTheme is dark or light
    action.setIcon(...);
});
```

Reply:
In addition, Firefox and Safari support setting context-menu item icon (See #592), and there is a proposal (#550) for setting sidebar icon. So besides `action.setIcon()` for the toolbar icon, there are other places where developers might need to dynamically set icons for dark/light theme in the future. So, in my option, supporting `runtime.getBrowserTheme()` and `runtime.onBrowserThemeChanged` might be a more general and easy-to-implement solution for dynamical icons.

Reply:
I do not support an event for this. The browser needs the color scheme icons before the user switches modes, otherwise there will be a flash of the other icons as we wait for the extension to respond to the event.

`menus.create()` and `menus.update()` should just support light/dark mode variants too.

Reply:
> I do not support a change event for this. The browser needs the color scheme icons before the user switches modes, otherwise there will be stale icons as we wait for the extension to respond to the event.

This is an important point that affects the design of the API. I'd like to know other browsers' opinions.

Reply:
I think there have been sufficient ideas shared to come to an understanding, represented in the latest patch.

Reply:
For me, as a developer, I don't think the change event listener or delaying some milliseconds is a problem. Instead, the other way requires developers to generate two icons every time and the browser need to hold two icons in memory, but only use one of them (approximately 4 KB of memory is wasted).

Anyway, I don't mind how it's implemented for setIcon API, as long as the functionality is there.

Reply:
@hanguokai Also maintaining extensions which use `imageData` for `setIcon`. Especially with different densities and the lack of figuring out what the current dark mode and devicePixelRatio is in serviceWorkers. The amount of variants could add up especially when different tabs are involved.

If we would provide this info to extension developers, I believe it should be on the action namespace and include more data. Something like:

`browser.action.getIconVariables()` which would give an object with data about the current conditions:
```json
{
  "color_scheme": "dark",
  "size": 24
}
```

And optionally
`browser.action.onIconVariablesChanged.addListener((newIconVariables) => { });`

Knowing the size the action icon will be used seems like a much bigger win than knowing the `color_scheme` as it is more stable and will provide more memory gains and more crispy icons when `imageData` is being used.